### PR TITLE
[v7r0] Changes and fixes for pilot wrappers

### PR DESCRIPTION
**Pull request by fstagni**
BEGINRELEASENOTES

*WMS
NEW: PilotWrapper can get the Pilot files from a list of locations
CHANGE: PilotWrapper is compatible with several python versions

*test
NEW: added a test for PilotWrapper

*docs
CHANGE: Pilot 3 is the default

ENDRELEASENOTES

### CODE REVIEW

File: WorkloadManagementSystem/Utilities/PilotWrapper.py [hunk 1]
```diff
@@ -144,20 +146,49 @@ def pilotWrapperScript(pilotFilesCompressedEncodedDict=None,
...
+    else:
+      rJson = urllib2.urlopen('https://' + loc + '/pilot/pilot.json',
+                              timeout=10)
+      rTar = urllib2.urlopen('https://' + loc + '/pilot/pilot.tar',
+                              timeout=10)
+      break
+
+  except urllib2.URLError:
+    print('%%s unreacheable' %% loc)
+
```
Comment (andresailer) [inline WorkloadManagementSystem/Utilities/PilotWrapper.py]: And an `else` case to for loop to throw an error if none of the urls worked?

File: docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst [hunk 2]
```diff
@@ -16,17 +16,16 @@ Bootstrap
 Pilot3 needs a JSON file for bootstrapping. We simply call this file the ``pilot.json`` file.
 The ``pilot.json`` file is created starting from information found in the Configuration Service.
 
-The pilot wrapper (the script that starts the pilot, which is effectively equivalent to what SiteDirectors send) expects
-to find (download) such a ``pilot.json`` file from a known location, which can be for example exposed by the DIRAC
-WebApp webserver.
+The pilot wrapper (the script that starts the pilot, which is effectively equivalent to what SiteDirectors send)
+expects to find (download) such pilot.json file from a known location, or a set of them.
```
Comment (andresailer) [inline docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst]:
```suggestion
expects to find and download such a ``pilot.json`` file from a known location, or a set of them.
```
if you just put download in parenthesis, then it should be "find ... at a known location"

File: docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst [hunk 3]
```diff
@@ -16,17 +16,16 @@ Bootstrap
...
+Such location can be, for example, exposed via *https://* by the DIRAC WebApp webserver. Other protocols (including *file://*) are possible.
```
Comment (andresailer) [inline docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst]:
```suggestion
Such a location can be, for example, exposed via *https://* by the DIRAC WebApp webserver. Other protocols (including *file://*) are possible.
```
I find this statement about the file protocol confusing. At the moment the provided pilot wrapper only takes a list of hosts, and then expects the files to be on that host, reachable via https and at a specific path. Or am I missing something?

File: docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst [hunk 4]
```diff
@@ -16,17 +16,16 @@ Bootstrap
...
 
-The ``pilot.json`` file is therefore always kept in sync with the content of the Configuration Service.
-From DIRAC v6r20, there is the possibility to set the option *UpdatePilotCStoJSONFile* to True in the configuration of
-the Configuration/Server service (please see :ref:`ConfigurationServer` for detais). If this option is set,
-at every configuration update, the ``pilot.json`` file content will also be updated (if necessary).
+The pilot.json file is therefore always kept in sync with the content of the Configuration Service.
```
Comment (andresailer) [inline docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst]:
```suggestion
The pilot.json file is always kept in sync with the content of the Configuration Service.
```
There was nothing written before about keeping the content in sync, so "therefore" is not applicable.

File: docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst [hunk 5]
```diff
@@ -16,17 +16,16 @@ Bootstrap
...
+At every configuration update, the pilot.json file content will also be updated (if necessary).
```
Comment (andresailer) [inline docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst]:
```suggestion
At every configuration update, the pilot.json file content will also be updated.
```

File: WorkloadManagementSystem/Utilities/PilotWrapper.py [hunk 6]
```diff
@@ -46,11 +47,11 @@
 logger.addHandler(screen_handler)
 
 # just logging the environment as first thing
-print '==========================================================='
+print('===========================================================')
```
Comment (andresailer) [inline WorkloadManagementSystem/Utilities/PilotWrapper.py]: Why not use logger?

File: WorkloadManagementSystem/Utilities/PilotWrapper.py [hunk 7]
```diff
@@ -144,20 +145,58 @@ def pilotWrapperScript(pilotFilesCompressedEncodedDict=None,
...
+
+    else:
+      rJson = url_library_urlopen('https://' + loc + '/pilot/pilot.json',
+                                  timeout=10)
+      rTar = url_library_urlopen('https://' + loc + '/pilot/pilot.tar',
+                                 timeout=10)
+      break
+
+  except url_library_URLError:
+    print('%%s unreacheable' %% loc)
```
Comment (andresailer) [inline WorkloadManagementSystem/Utilities/PilotWrapper.py]: Also log the message of the exception, also logger?

File: docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst [hunk 8]
```diff
@@ -16,17 +16,16 @@ Bootstrap
...
 
-If *UpdatePilotCStoJSONFile* is True, then also the Operations option *Pilot/<...>/pilotFileServer* should be set to the webserver chosen for the upload.
-We suggest to use simply the DIRAC webserver.
+Also the Operations option *Pilot/<...>/pilotFileServer* should be set to the webserver(s) chosen for the upload.
+If more than one location are used, add them as a list.
```
Comment (andresailer) [inline docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst]:
```suggestion
If more than one location is used, add them as a list.
```

File: docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst (hunk 3 above)
Comment (fstagni) [reply docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst]: The pilot wrapper may find such file(s) in the local file system, e.g. on CVMFS, so `file:///cvmfs/some/path.json` (I think `urllib` and `request` handles these cases).

Comment (andresailer) [reply docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst]: And how is the pilotWrapper told to look at, e.g., cvmfs? Looking at the pilotWrapper code the only urlopen calls explicitly contain `'https://' + loc`

Comment (fstagni) [reply docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst]: Ah, you are right, I was mixing with my running test. Now, 2 options: 
- change the doc and leave the cod as is
- change the code
is the second option needed (now)?

Comment (andresailer) [reply docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst]: Change the doc.

Comment (fstagni) [reply docs/source/AdministratorGuide/Systems/WorkloadManagement/Pilots/Pilots3.rst]: Done.

**Pull request merged**